File: cole/day13_diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def d(self, root: Optional[TreeNode]) -> tuple[int]:
        if not root:
            return 0, 0
        ld, lP = self.d(root.left)
        rd, rP = self.d(root.right)
        d = max(ld, rd, lP + rP)
        P = max(lP, rP) + 1
        return d, P

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return self.d(root)[0]

File: cole/day13_diameter-of-binary-tree/diameter-of-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#define MAX(a, b) (a > b ? a : b)
int *d(struct TreeNode *root, int dP[2])
{
    if (!root)
        return dP;
    int l[] = {0, 0}, r[] = {0, 0};
    d(root->left, l);
    d(root->right, r);
    dP[0] = MAX(MAX(l[0], r[0]), l[1] + r[1]);
    dP[1] = MAX(l[1], r[1]) + 1;
    return dP;
}

int diameterOfBinaryTree(struct TreeNode *root)
{
    int dP[] = {0, 0};
    return d(root, dP)[0];
}